File: shared/csharp/src/Authorizer.cs

using System.Collections.Generic;
using PebbleFields.PebbleObject;
namespace PebbleFields.Shared
{

	/**
	 * Helper class you can inherit from when writing authorizers
	 * @author rtsunoda
	 *
	 */
	public abstract class Authorizer {

		public static readonly string UNAUTHORIZED_USER = "-1";
		protected string userId = UNAUTHORIZED_USER;

		public Authorizer() 
		{

		}

		public Pebble getUserObject() 
		{
			Pebble user = new Pebble();

			Pebble top = DocModel.getDoc("theInstance");//get top app
			List<Pebble> maps = top.getRecords("groupMaps");
			//check against all groupMaps if true for any then authorized
			foreach (Pebble map in maps) 
			{
				string actorGroup = map.getRef(".");
				Pebble userGroup = map.get("toGroup");
				if (this.isInGroup(userGroup)) 
				{
					if (actorGroup.IndexOf(".") != -1) 
					{
						actorGroup = actorGroup.Substring(actorGroup.LastIndexOf(".") + 1);
					}
					user.set("roles." + actorGroup, new Pebble());
				}
			}
			user.setRef("id", this.userId);
			bool isAuth = this.userId != UNAUTHORIZED_USER;
			user.setValue("isAuth", isAuth.ToString());
			return user;
		}
		abstract public bool isInGroup(Pebble userGroup);
	}
}

File: shared/javascript/src/authorizer.js

goog.provide("pebble.shared.Authorizer");

goog.require("pebble.Pebble");
goog.require("pebble.shared.DocModel");
//-----file start

/**
 * Helper class you can inherit from when writing authorizers
 * @constructor
 */
pebble.shared.Authorizer = function(userId) {
	this.userId = userId || pebble.shared.Authorizer.UNAUTHORIZED_USER;
}
/**
 * @return {pebble.Pebble}
 */
pebble.shared.Authorizer.prototype.getUserObject = function() {
	var user = new pebble.Pebble();

	var top = pebble.shared.DocModel.getDoc("theInstance");//get top app
	var maps = top.getRecords("groupMaps");
	//check against all groupMaps if true for any then authorized
	for (var j = 0; j < maps.length; j++) {
		var map = maps[j];
		var actorGroup = map.getRef(".");
		var userGroup = map.get("toGroup");
		if (this.isInGroup(actorGroup, userGroup)) {
			user.set("roles." + actorGroup, new pebble.Pebble());
		}
	}
	user.setRef("id", this.userId);
	var isAuth = this.userId != pebble.shared.Authorizer.UNAUTHORIZED_USER;
	user.setValue("isAuth", isAuth.toString());
	return user;
};
/**
 * @param {pebble.Pebble} userGroup
 * @return {boolean}
 */
pebble.shared.Authorizer.prototype.isInGroup = function(userGroup){}


pebble.shared.Authorizer.UNAUTHORIZED_USER = "-1";
